### ISSUE TITLE
Out of memory (server) and slow interface

### ISSUE THREAD

**Issue by mdbenito** [vivekratnavel/omniboard#87]
Hi, I'm running Omniboard 1.4.0 from the docker image using docker-compose and it's being very useful, thanks!

However, with a medium database (~2K objects) and default memory settings for node.js omniboard sometimes fails to even load. Each run contains many metrics and artifacts and is pretty heavy.

"Sometimes failing" means that from a fresh tab and freshly started server, I have out of memory on the server and an unresponsive browser. And if I managed to apply some filters while the database was still smaller, then sometimes I see results in the browser, but the server constantly crashes.

The container's logs typically look like this:
```
omniboard_1  | ::ffff:172.25.0.1 - - [07/Apr/2019:09:08:47 +0000] "GET /static/js/main.f3c8819a.js.map HTTP/1.1" 200 28763040 7.656 ms             omniboard_1  |                                                                                                                                     omniboard_1  | <--- Last few GCs --->                                                                                                              omniboard_1  |                                                                                                                                     omniboard_1  | [25:0x2b1bab0]    24199 ms: Mark-sweep 1209.6 (1270.3) -> 1209.6 (1239.3) MB, 530.6 / 0.0 ms  (average mu = 0.677, current mu = 0.000) last resort GC in old space requested                                                                                                           omniboard_1  | [25:0x2b1bab0]    24721 ms: Mark-sweep 1209.6 (1239.3) -> 1209.6 (1239.3) MB, 522.1 / 0.0 ms  (average mu = 0.500, current mu = 0.000) last resort GC in old space requested                                                                                                           omniboard_1  |                                                                                                                                     omniboard_1  |                                                                                                                                     omniboard_1  | <--- JS stacktrace --->                                                                                                             omniboard_1  |                                                                                                                                     
omniboard_1  | ==== JS stack trace =========================================                                                                       omniboard_1  |                                                                                                                                     omniboard_1  |     0: ExitFrame [pc: 0x330cbc1041bd]                                                                                               
omniboard_1  | Security context: 0x6733981e789 <JSObject>                                                                                          omniboard_1  |     1: byteLength(aka byteLength) [0x2dd4e6d0ffb9] [buffer.js:530] [bytecode=0x31ea937719d1 offset=204](this=0x2d4e972822e1 <undefined>,string=0xccc011a74d1 <Very long string[301641766]>,encoding=0x67339831cc9 <String[4]: utf8>)                                                   
omniboard_1  |     2: arguments adaptor frame: 3->2                                                                                                omniboard_1  |     3: fromString(aka fromString) [0x2dd4e6d27009] [buffer.js:341] [bytecode=0x1b26a7e7259 offs...                                  omniboard_1  |                                                                                                                                     
omniboard_1  | FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory                                                  
omniboard_1  |  1: 0x8b8210 node::Abort() [node]                                                                                                   
omniboard_1  |  2: 0x8b825c  [node]                                                                                                                
omniboard_1  |  3: 0xac1d1e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]                                          omniboard_1  |  4: 0xac1f38 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]                            
omniboard_1  |  5: 0xeb11f2  [node]                                                                                                                
omniboard_1  |  6: 0xec0466 v8::internal::Heap::AllocateRawWithRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node] 
omniboard_1  |  7: 0xe8e86e v8::internal::Factory::NewRawOneByteString(int, v8::internal::PretenureFlag) [node]                                    
omniboard_1  |  8: 0xfd3b2d v8::internal::String::SlowFlatten(v8::internal::Handle<v8::internal::ConsString>, v8::internal::PretenureFlag) [node]  
omniboard_1  |  9: 0xabf454 v8::internal::String::Flatten(v8::internal::Handle<v8::internal::String>, v8::internal::PretenureFlag) [node]          
omniboard_1  | 10: 0xacd79f v8::String::Utf8Length() const [node]                                                                                  
omniboard_1  | 11: 0x8d0ed2  [node]                                                                                                                omniboard_1  | 12: 0xb4ee49  [node]                                                                                                                omniboard_1  | 13: 0xb4f9b9 v8::internal::Builtin_HandleApiCall(int, v8::internal::Object**, v8::internal::Isolate*) [node]                        
omniboard_1  | 14: 0x330cbc1041bd                                                                                                                  
omniboard_1  | npm ERR! code ELIFECYCLE                                                                                                            
omniboard_1  | npm ERR! errno 1                                                                                                                    
omniboard_1  | npm ERR! omniboard@1.4.0 prod: `cross-env NODE_ENV=production node dist/index.js "-m" "mongo:27017:lvk"`                            omniboard_1  | npm ERR! Exit status 1                                                                                                              omniboard_1  | npm ERR!                                                                                                                            
omniboard_1  | npm ERR! Failed at the omniboard@1.4.0 prod script.                                                                                 
omniboard_1  | npm ERR! This is probably not a problem with npm. There is likely additional logging output above.                                  
omniboard_1  |                                                                                                                                     
omniboard_1  | npm ERR! A complete log of this run can be found in:                                                                                omniboard_1  | npm ERR!     /root/.npm/_logs/2019-04-07T09_09_34_844Z-debug.log                                                                   omniboard_1  |                                                                                                                                     
omniboard_1  | > omniboard@1.4.0 prod /usr/omniboard                                                                                               
omniboard_1  | > cross-env NODE_ENV=production node dist/index.js "-m" "mongo:27017:lvk"                                                           
omniboard_1  |                                                                                                                                     
omniboard_1  | Omniboard is listening on port 9000!                                       
```

The referred npm log is

```
root@eebf01b8a16b:~/.npm/_logs# cat 2019-04-07T09_09_34_844Z-debug.log                                                                            
0 info it worked if it ends with ok                                                                                                                
1 verbose cli [ '/usr/local/bin/node',                                                                                                            
1 verbose cli   '/usr/omniboard/node_modules/.bin/npm',                                                                                            
1 verbose cli   'run',                                                                                                                            
1 verbose cli   'prod',                                                                                                                            
1 verbose cli   '--',                                                                                                                              
1 verbose cli   '-m',                                                                                                                              
1 verbose cli   'mongo:27017:lvk' ]                                                                                                                
2 info using npm@6.5.0                                                                                                                             
3 info using node@v10.8.0                                                                                                                          
4 verbose run-script [ 'preprod', 'prod', 'postprod' ]                                                                                             
5 info lifecycle omniboard@1.4.0~preprod: omniboard@1.4.0                                                                                         
6 info lifecycle omniboard@1.4.0~prod: omniboard@1.4.0                                                                                             
7 verbose lifecycle omniboard@1.4.0~prod: unsafe-perm in lifecycle true                                                                            
8 verbose lifecycle omniboard@1.4.0~prod: PATH: /usr/omniboard/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/usr/omniboard/node_module$
/.bin:/usr/omniboard/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin                                                
9 verbose lifecycle omniboard@1.4.0~prod: CWD: /usr/omniboard                                                                                      
10 silly lifecycle omniboard@1.4.0~prod: Args: [ '-c',                                                                                             
10 silly lifecycle   'cross-env NODE_ENV=production node dist/index.js "-m" "mongo:27017:lvk"' ]                                                   
11 silly lifecycle omniboard@1.4.0~prod: Returned: code: 1  signal: null                                                                           
12 info lifecycle omniboard@1.4.0~prod: Failed to exec prod script                                                                                 
13 verbose stack Error: omniboard@1.4.0 prod: `cross-env NODE_ENV=production node dist/index.js "-m" "mongo:27017:lvk"`                            
13 verbose stack Exit status 1                                                                                                                     
13 verbose stack     at EventEmitter.<anonymous> (/usr/omniboard/node_modules/npm/node_modules/npm-lifecycle/index.js:301:16)                      
13 verbose stack     at EventEmitter.emit (events.js:182:13)                                                                                       
13 verbose stack     at ChildProcess.<anonymous> (/usr/omniboard/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)                   
13 verbose stack     at ChildProcess.emit (events.js:182:13)                                                                                       
13 verbose stack     at maybeClose (internal/child_process.js:962:16)                                                                              
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:249:5)                                                      
14 verbose pkgid omniboard@1.4.0                                                                                                                   
15 verbose cwd /usr/omniboard                                                                                                                      
16 verbose Linux 4.13.0-46-generic                                                                                                                 
17 verbose argv "/usr/local/bin/node" "/usr/omniboard/node_modules/.bin/npm" "run" "prod" "--" "-m" "mongo:27017:lvk"                             
18 verbose node v10.8.0                                                                                                                           
19 verbose npm  v6.5.0                                                                                                                             
20 error code ELIFECYCLE
21 error errno 1
22 error omniboard@1.4.0 prod: `cross-env NODE_ENV=production node dist/index.js "-m" "mongo:27017:lvk"`
22 error Exit status 1
23 error Failed at the omniboard@1.4.0 prod script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

If I up the memory limit with `NODE_OPTIONS=--max-old-space-size=8192` the server **doesn't crash anymore**, but the browser's (chrome 73) memory usage goes to around 1.5GB and becomes totally unresponsive.

If I try to apply filters or deselect some columns, the UI seems not to react, but after a reload my changes have been applied and I can work. I know that some of the loading is done only when unfolding individual experiments (issue #47 ) but that solution seems not to be enough (?)

A couple of suggestions for workarounds:

* Paging! infinite scrolling is nice, but are the queries paged too?
* Make the checkboxes in the dropdown "show/hide columns" **not** trigger redraws before the dropdown closes.
* Do not try to compute valid filter values for numeric columns (feels like it is being done for any column).
* More indices??

Thanks for any help.

**Comment by vivekratnavel** [comment]
Hi @mdbenito, 

Thanks for the detailed logs. I will definitely take a look at this. 

I think the best way to tackle this issue is to have support for pagination. I will look into the feasibility of supporting pagination. Also, is it possible for you to send a dump of your database after stripping any sensitive data to my email? It would greatly help with testing performance quickly.

**Comment by vivekratnavel** [comment]
This will be fixed in v2.0 #107 

Filtering and sorting is moved to server side in v2.0 and this improves performance significantly.

**Comment by vivekratnavel** [comment]
Thanks @mdbenito for sharing your sacred database with me. It helped a lot in testing performance of Omniboard and paved a way for v2.0.0.

Omniboard v2.0.0-beta is out. You can try it out with either npm or docker - https://www.npmjs.com/package/omniboard/v/2.0.0-beta.2 
Docker image - vivekratnavel/omniboard:beta
Source code is in branch 'beta' - https://github.com/vivekratnavel/omniboard/tree/beta

Please try it out and give feedback.
